File: bender/app/src/main/java/kz/app/bender/mvp/tabsNavigation/FlowFragmentMain.java
package kz.app.bender.mvp.tabsNavigation;

import android.support.annotation.NonNull;
import android.support.design.widget.BottomNavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;

import com.arellomobile.mvp.presenter.InjectPresenter;
import com.arellomobile.mvp.presenter.ProvidePresenter;

import butterknife.BindView;
import kz.app.bender.R;
import kz.app.bender.mvp.AbstractView;
import kz.app.bender.mvp.IndigoMVPFragment;
import kz.app.bender.support.navigation.BenderRouter;
import kz.app.bender.support.navigation.RouterProvider;
import kz.app.bender.support.navigation.Screens;
import kz.app.bender.support.provider.FragmentProvider;
import ru.terrakok.cicerone.Cicerone;
import ru.terrakok.cicerone.Navigator;
import ru.terrakok.cicerone.android.support.SupportAppNavigator;
import ru.terrakok.cicerone.commands.Command;

public class FlowFragmentMain extends IndigoMVPFragment<FlowFragmentMainPresenter>
        implements RouterProvider, FlowFragmentContract.View {

    //region Views
    @BindView(R.id.bottomNavigationView)
    BottomNavigationView bottomNavigationView;
    //endregion

    //region Vars
    private Cicerone<BenderRouter> tabsRouter;
    private Navigator tabsNavigator;
    //endregion

    //region DI
    @InjectPresenter
    FlowFragmentMainPresenter mPresenter;

    @ProvidePresenter
    FlowFragmentMainPresenter createFlowFragmentMainPresenter() {
        createTabsRouterIfNecessary();
        return new FlowFragmentMainPresenter(getRouter(), tabsRouter.getRouter());
    }
    //endregion

    //region New instance
    public static FlowFragmentMain newInstance() {
        return new FlowFragmentMain();
    }
    //endregion

    //region Overridden methods
    @NonNull
    @Override
    public FragmentProvider getFragmentProvider() {
        return new FragmentProvider().setLayoutResourceId(R.layout.fragment_tabs_navigation);
    }

    @NonNull
    public FlowFragmentMainPresenter getPresenter() {
        return mPresenter;
    }

    @Override
    public BenderRouter provideRouter() {
        return tabsRouter.getRouter();
    }

    @Override
    public void onResume() {
        createTabsRouterIfNecessary();
        super.onResume();
        mPresenter.setTabsRouter(tabsRouter.getRouter());
    }

    @Override
    public void onPause() {
        tabsRouter.getNavigatorHolder().removeNavigator();
        super.onPause();
    }
    //endregion

    //region UI
    @Override
    public void prepareUI() {
        bottomNavigationView.setOnNavigationItemSelectedListener(menuItem -> {
            switch (menuItem.getItemId()) {
                case R.id.bottom_bar_main_tab:
                    mPresenter.onTabClicked(Screens.Tabs.tab_login);
                    return true;
                case R.id.bottom_bar_web_site_karlygash:
                    mPresenter.onTabClicked(Screens.Tabs.tab_web_site_karlygash);
                    return true;
                case R.id.bottom_bar_web_site_profile:
                    mPresenter.onTabClicked(Screens.Tabs.tab_web_site_profile);
                    return true;
            }
            return false;
        });
        bottomNavigationView.setSelectedItemId(R.id.bottom_bar_main_tab);
    }
    //endregion

    //region Support
    private void createTabsRouterIfNecessary() {
        if (tabsRouter == null)
            tabsRouter = Cicerone.create(new BenderRouter());

        if (tabsNavigator == null)
            tabsNavigator = new SupportAppNavigator(activity, getChildFragmentManager(), R.id.flFragmentFrame) {
                @Override
                protected void setupFragmentTransaction(Command command, Fragment currentFragment, Fragment nextFragment, FragmentTransaction fragmentTransaction) {
                    fragmentTransaction.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN);
                }
            };
        tabsRouter.getNavigatorHolder().setNavigator(tabsNavigator);
    }
    //endregion
}

File: DeliveryApp/app/src/main/java/kz/app/deliveryapp/mvp/tabsNavigation/FlowFragmentMain.kt
package kz.app.deliveryapp.mvp.tabsNavigation

import android.support.design.widget.NavigationView
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentTransaction
import android.view.Gravity
import com.arellomobile.mvp.presenter.InjectPresenter
import com.arellomobile.mvp.presenter.ProvidePresenter
import kotlinx.android.synthetic.main.fragment_navigation_drawer.*
import kz.app.deliveryapp.R
import kz.app.deliveryapp.mvp.RootMvpAppCompatFragment
import kz.app.deliveryapp.mvp.основнойКонтейнер.RootPresenter
import kz.app.deliveryapp.support.навигация.BenderRouter
import kz.app.deliveryapp.support.навигация.RouterProvider
import kz.app.deliveryapp.support.навигация.Screens
import kz.app.deliveryapp.support.провайдерДляТулбара.FragmentProvider
import ru.terrakok.cicerone.Cicerone
import ru.terrakok.cicerone.Navigator
import ru.terrakok.cicerone.android.support.SupportAppNavigator
import ru.terrakok.cicerone.commands.Command

class FlowFragmentMain : RootMvpAppCompatFragment<FlowFragmentMainPresenter>(), RouterProvider,
    FlowFragmentContract.View {

    //region Vars
    private var tabsRouter: Cicerone<BenderRouter>? = null
    private var tabsNavigator: Navigator? = null
    //endregion

    //region DI
    @InjectPresenter
    lateinit var mPresenter: FlowFragmentMainPresenter

    @ProvidePresenter
    internal fun createFlowFragmentMainPresenter(): FlowFragmentMainPresenter {
        createTabsRouterIfNecessary()
        return FlowFragmentMainPresenter(router, tabsRouter!!.router)
    }
    //endregion

    //region Overridden methods
    override fun getFragmentProvider(): FragmentProvider {
        return FragmentProvider().setLayoutResourceId(R.layout.fragment_navigation_drawer)
            .setToolbarTitle(getString(R.string.text_menu), R.color.colorAccent)
            .setToolbarHamburgerColor(R.color.colorAccent)
    }

    override fun getPresenter(): FlowFragmentMainPresenter {
        return mPresenter
    }

    override fun provideRouter(): BenderRouter {
        return tabsRouter!!.router
    }

    override fun onResume() {
        createTabsRouterIfNecessary()
        super.onResume()
        mPresenter.setTabsRouter(tabsRouter!!.router)
    }

    override fun onPause() {
        tabsRouter!!.navigatorHolder.removeNavigator()
        super.onPause()
    }
    //endregion

    //region UI
    override fun prepareUI() {
        if (fragmentProvider.toolbarHamburgerColor != null)
            toolbar?.setNavigationOnClickListener { drawer_layout.openDrawer(Gravity.START) }

        navigation.setNavigationItemSelectedListener(NavigationView.OnNavigationItemSelectedListener { item ->
            when (item.itemId) {
                R.id.nav_menu_menu -> {
                    toolbar?.setTitle(R.string.text_menu)
                    mPresenter.onNavTabClicked(Screens.Tabs.nav_tab_menu)
                }
                R.id.nav_menu_basket -> {
                    RootPresenter.instance.flowNavigateTo(Screens.getBasketList())
                }
                R.id.nav_menu_aksiya -> {
                    toolbar?.setTitle(R.string.text_aksiya)
                    mPresenter.onNavTabClicked(Screens.Tabs.nav_tab_aksii)
                }
                R.id.nav_menu_news -> {
                    toolbar?.setTitle(R.string.text_news)
                    mPresenter.onNavTabClicked(Screens.Tabs.nav_tab_news)
                }
                else -> return@OnNavigationItemSelectedListener true
            }
            drawer_layout.closeDrawer(Gravity.START)
            true
        })
    }
    //endregion

    //region Support
    private fun createTabsRouterIfNecessary() {
        if (tabsRouter == null)
            tabsRouter = Cicerone.create(BenderRouter())

        if (tabsNavigator == null)
            tabsNavigator = object : SupportAppNavigator(activity, childFragmentManager, R.id.flFragmentFrameDrawer) {
                override fun setupFragmentTransaction(
                    command: Command?,
                    currentFragment: Fragment?,
                    nextFragment: Fragment?,
                    fragmentTransaction: FragmentTransaction?
                ) {
                    fragmentTransaction!!.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                }
            }
        tabsRouter!!.navigatorHolder.setNavigator(tabsNavigator)
    }

    companion object {
        fun newInstance(): FlowFragmentMain {
            return FlowFragmentMain()
        }
    }
}
